net: lib: nrf_cloud: use new c2d topics

jayteemo opened the pull request:
Use new topics to receive data from nRF Cloud.
Data for location services now comes in on separate topics to allow for better filtering.
A-GPS and P-GPS is no longer forwarded to the application. Cellular positioning data is forwarded only if a callback has not been registered.
Created new event IDs for RX data.
NCSDK-11891

Signed-off-by: Justin Morton <justin.morton@nordicsemi.no>

NordicBuilder commented:
You can find the documentation preview for this PR at this <a href='http://developer.nordicsemi.com/nRF_Connect_SDK_dev/doc/PR-8780
'>link</a>. It will be updated about 10 minutes after the documentation build succeeds.

*Note: This comment is automatically posted by the Documentation Publishing GitHub Action.*

jayteemo commented:
draft until cloud changes are live on prod.
testable on dev.

plskeggs commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:
```diff
@@ -530,19 +541,76 @@ static int cell_pos_cb_send(const char *const rx_buf)
 		.data = nct_evt->param.dc->data,
 		.topic = nct_evt->param.dc->topic,
 	};
 
```
It would be nice to have the topic logged here, ideally with the data, though that's problematic for A-GPS.

plskeggs commented:
Looks good so far.  Certainly simplifies the integration of A-GPS and P-GPS.

glarsennordic reviewed (commented):
I like it so far too!

jayteemo commented:
Note: this PR is "complete", i'm not making any changes other than rebasing and addressing PR comments.
it is only in draft because the cloud changes are not on prod.

jayteemo commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:
```diff
@@ -894,9 +894,10 @@ static void nct_mqtt_evt_handler(struct mqtt_client *const mqtt_client,
 		const struct mqtt_publish_param *p = &_mqtt_evt->param.publish;
 
-		LOG_DBG("MQTT_EVT_PUBLISH: id = %d len = %d",
+		LOG_DBG("MQTT_EVT_PUBLISH: id = %d len = %d, topic = %s",
```
Added [requested topic logging](https://github.com/nrfconnect/sdk-nrf/pull/8780#discussion_r974716168) here @plskeggs

glarsennordic commented on applications/asset_tracker_v2/src/modules/cloud_module.c:
```diff
@@ -270,48 +270,60 @@ static bool app_event_handler(const struct app_event_header *aeh)
+	} else if (err == -ENODATA) {
+		LOG_WRN("Device configuration empty!");
+		SEND_EVENT(cloud, CLOUD_EVT_CONFIG_EMPTY);
+	} else if (err == -ECANCELED) {
+		/* The incoming message has already been handled, ignored. */
+	} else if (err == -ENOENT) {
+		/* Encoding of incoming message is not supported. */
 	} else {
```
Debug logs for these could be useful

glarsennordic commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_codec.c:
```diff
@@ -688,12 +712,12 @@ int nrf_cloud_decode_data_endpoint(const struct nrf_cloud_data *input,
+				    rx_endpoint);
 	if (err) {
 		cJSON_Delete(root_obj);
-		LOG_ERR("could not decode topic for %s", JSON_KEY_CLOUD_TO_DEVICE);
```
Nit: Most of the logs in this file are capitalized

glarsennordic replied:
Consider capitalizing the "could not decode" occurance on 671 as well

glarsennordic commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:
```diff
@@ -530,19 +541,79 @@ static int cell_pos_cb_send(const char *const rx_buf)
+
+#if defined(CONFIG_NRF_CLOUD_PGPS)
+	if (ret == 0) {
+		ret = nrf_cloud_pgps_notify_prediction();
```
Remind me again why this must be called in response to agps data receipt? I don't question it, but I would like to fully understand

jayteemo replied:
i don't know, but i bet @plskeggs does

plskeggs replied:
If both A-GPS and P-GPS are enabled, everything but ephemerides and almanacs are handled by A-GPS.  In this configuration, it is important to check, after receiving A-GPS data, whether any further assistance is needed by the modem for ephemerides, which would come from P-GPS (usually, in stored predictions in flash).

jayteemo replied:
maybe that should be a comment in the code somewhere?  or maybe it already is?

glarsennordic commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:
```diff
@@ -530,19 +541,79 @@ static int cell_pos_cb_send(const char *const rx_buf)
+int cell_pos_process(const char * const buf)
+{
+#if defined(CONFIG_NRF_CLOUD_CELL_POS)
+	if (cell_pos_cb_send(buf) == 0) {
```
It seems like the contents of `cell_pos_cb_send` should just be moved into this function

glarsennordic commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_fsm.c:
```diff
@@ -530,19 +541,79 @@ static int cell_pos_cb_send(const char *const rx_buf)
+	case NRF_CLOUD_RCV_TOPIC_GENERAL:
+	default:
+		cloud_evt.type = NRF_CLOUD_EVT_RX_DATA_GENERAL;
+		discon_req = nrf_cloud_detect_disconnection_request(cloud_evt.data.ptr);
```
I feel that this should have its own topic (a device control topic), but given that that would require further BE changes, maybe that would be for a round-two PR

glarsennordic commented on subsys/net/lib/nrf_cloud/src/nrf_cloud_transport.c:
```diff
@@ -894,9 +894,10 @@ static void nct_mqtt_evt_handler(struct mqtt_client *const mqtt_client,
 			p->message_id,
-			p->message.payload.len);
+			p->message.payload.len,
+			p->message.topic.topic.utf8);
```
This will be nice to have, though I'm not sure whether name is guaranteed to be null-terminated. You may want to use `%.*s` with `message.topic.topic.size`. Or do you know for sure it is null-terminated?

jayteemo commented:
added a commit for error events based on a comment from @simensrostad

jayteemo commented on applications/asset_tracker_v2/src/modules/cloud_module.c:
```diff
@@ -270,48 +270,60 @@ static bool app_event_handler(const struct app_event_header *aeh)
+	} else if (err == -ENODATA) {
+		LOG_WRN("Device configuration empty!");
+		SEND_EVENT(cloud, CLOUD_EVT_CONFIG_EMPTY);
+	} else if (err == -ECANCELED) {
+		/* The incoming message has already been handled, ignored. */
+	} else if (err == -ENOENT) {
+		/* Encoding of incoming message is not supported. */
 	} else {
```
code was just moved around for this change.  i'll leave that up to @simensrostad in a future atv2 update.

glarsennordic replied:
Ah you're right. I lost sight of the fact I was looking at ATV2 code not library code.

glarsennordic reviewed (approved):
LGTM

junqingzou commented on applications/serial_lte_modem/src/gnss/slm_at_gnss.c:
```diff
@@ -733,44 +733,41 @@ static void on_cloud_evt_disconnected(void)
-	}
+static void on_cloud_evt_data_received(const struct nrf_cloud_data *const data)
+{
+	sprintf(rsp_buf, "\r\n#XNRFCLOUD: %s\r\n", (char *)data->ptr);
```
Need below condition be satisfied

if (nrf_cloud_ready) {

jayteemo replied:
i don't think that is needed anymore, but i will add it.
this is the handler for the new event `NRF_CLOUD_EVT_RX_DATA_GENERAL`; there should not be any of these events before the cloud is ready.
a new event `NRF_CLOUD_EVT_RX_DATA_SHADOW` may occur before the cloud is ready, but this data is ignored by slm.

junqingzou replied:
Yes it was the shadow data that confused customers before. Logically we should only allow uplink/downlink "general" after `NRF_CLOUD_EVT_READY`, so please add the `if` check even it's redundant now.

junqingzou commented:
Is it correct understanding that APP is still responsible to make A-GPS and P-GPS request to nRF Cloud (by calling lib_nrf_cloud API), but the response is now all handled in side the lib_nrf_cloud? Will lib_nrf_cloud send back an event with the result of handling A-GPS and P-GPS request?

jayteemo commented:
> Will lib_nrf_cloud send back an event with the result of handling A-GPS and P-GPS request?

Only on a processing error.
Is there a use case for the application to receive an event indicating that A/P-GPS data has been processed?

junqingzou commented:
> > Will lib_nrf_cloud send back an event with the result of handling A-GPS and P-GPS request?
> 
> Only on a processing error. Is there a use case for the application to receive an event indicating that A/P-GPS data has been processed?

At least currently not required for SLM application, just for the sake of request-response model.

jayteemo commented on include/net/nrf_cloud.h:
```diff
@@ -111,10 +117,37 @@ enum nrf_cloud_evt_type {
+	/** MQTT service is unavailable */
+	NRF_CLOUD_ERR_STATUS_MQTT_CONN_SERVER_UNAVAIL,
+	/** Malformed user name or password */
+	NRF_CLOUD_ERR_STATUS_MQTT_CONN_BAD_USR_PWD,
```
i was copying the general format of the source error enum:
https://github.com/nrfconnect/sdk-zephyr/blob/97c4a1a6d1bda83f0272d9bbf68b211efadc7b4a/include/zephyr/net/mqtt.h#L130

junqingzou reviewed (approved):
Changes in SLM/GNSS are good

jayteemo commented:
Removing draft status. The required cloud BE changes to prod are about a week away.  Labeling with `DNM` until then.

NordicBuilder commented:
### Integration test specification

| Test Module | File based changes | Manually selected | West overwrite |
| ----------- | :----------------: | :---------------: | :------------: |
| test-ci-nrfconnect-boot-fw-update | X | | |
| test-fw-nrfconnect-chip | X | | |
| test-fw-nrfconnect-nrf-iot_mosh | X | | |
| test-fw-nrfconnect-nrf-iot_positioning | X | | |
| test-fw-nrfconnect-nrf-iot_thingy91 | X | | |
| test-sdk-homekit | X | | |


[Detailed information of selected test modules](https://jenkins-ncs.nordicsemi.no/job/latest/job/sub/job/test-fw-nrfconnect-nrf/job/master/27570/artifact/test-spec/test-spec-summary.yml/*view*/)

*Note: This message is automatically posted and updated by the CI*

trantanen commented:
When prod cloud is updated to support these changes, will it continue to handle older requests/responses from older implementations?

trantanen commented on applications/asset_tracker_v2/src/cloud/nrf_cloud_integration.c:
```diff
@@ -228,24 +222,20 @@ static void nrf_cloud_event_handler(const struct nrf_cloud_evt *evt)
 		cloud_wrap_evt.data.len = evt->data.len;
 		notify = true;
 		break;
+	case NRF_CLOUD_EVT_RX_DATA_CELL_POS:
+		LOG_DBG("NRF_CLOUD_EVT_RX_DATA_CELL_POS");
```
I have another PR where I'm reading cell pos results in ATv2. I guess I can find them still at `evt->data.ptr`, right?

jayteemo replied:
correct

tokangas commented on subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h:
```diff
@@ -128,6 +128,15 @@ extern "C" {
+	NRF_CLOUD_RCV_TOPIC_PGPS,
+	NRF_CLOUD_RCV_TOPIC_CELL_POS,
+	/* Unhandled or unable to be processed */
+	NRF_CLOUD_RCV_TOPIC_UNHANDLED
```
UNKNOWN sounds good to me.

tokangas reviewed (approved):
Looks good.

jayteemo commented:
> When prod cloud is updated to support these changes, will it continue to handle older requests/responses from older implementations?

correct

peknis commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -342,6 +370,12 @@ Libraries for networking
+    * Replaced event :c:enum:`NRF_CLOUD_EVT_RX_DATA` with :c:enum:`NRF_CLOUD_EVT_RX_DATA_GENERAL`.
+      Added events :c:enum:`NRF_CLOUD_EVT_RX_DATA_CELL_POS` and :c:enum:`NRF_CLOUD_EVT_RX_DATA_SHADOW`.
+    * The library now processes A-GPS and P-GPS data; it is no longer passed to the application.
+    * The status field of :c:enum:`NRF_CLOUD_EVT_ERROR` events use values from the enumeration nrf_cloud_error_status.
```
```suggestion
    * The status field of :c:enum:`NRF_CLOUD_EVT_ERROR` events use values from the enumeration :c:enumerator:`nrf_cloud_error_status`.
```

peknis commented on include/net/nrf_cloud.h:
```diff
@@ -144,7 +177,10 @@ enum nrf_cloud_connect_result {
 
-/**@ nRF Cloud error codes. */
+/**@ nRF Cloud error codes.
+ * See the Error Codes section of nRF Cloud API documentation for
```
```suggestion
 * See the <a href="https://api.nrfcloud.com/v1#section/Error-Codes">Error Codes</a> section of nRF Cloud API documentation for
```

peknis commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -342,6 +370,12 @@ Libraries for networking
+    * The library now subscribes to a wildcard cloud-to-device (c2d) topic.
+      This enables the device to receive location services data on separate topics.
+    * Replaced event :c:enum:`NRF_CLOUD_EVT_RX_DATA` with :c:enum:`NRF_CLOUD_EVT_RX_DATA_GENERAL`.
+      Added events :c:enum:`NRF_CLOUD_EVT_RX_DATA_CELL_POS` and :c:enum:`NRF_CLOUD_EVT_RX_DATA_SHADOW`.
```
```suggestion
     * Added events :c:enum:`NRF_CLOUD_EVT_RX_DATA_CELL_POS` and :c:enum:`NRF_CLOUD_EVT_RX_DATA_SHADOW`.
```

trantanen commented:
> Removing draft status. The required cloud BE changes to prod are about a week away. Labeling with `DNM` until then.

What is the expected timeline for this PR to be merged? That is, related cloud changes to be in prod?

peknis commented on subsys/net/lib/nrf_cloud/include/nrf_cloud_codec.h:
```diff
@@ -234,7 +244,7 @@ int json_send_to_cloud(cJSON * const request);
 cJSON *json_create_req_obj(const char *const app_id, const char *const msg_type);
 
-/** @brief Parses received REST data for an nRF Cloud error code  */
+/** @brief Parses received REST data for an nRF Cloud error code */
```
A general note on the verb form in all function descriptions. You should always use the imperative format: https://developer.nordicsemi.com/nRF_Connect_SDK/doc/latest/nrf/doc_styleguide.html#functions

This header file seems to have most of them in the third person format.

peknis reviewed (commented):
Left a general note on the function descriptions.

melwee01 commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -445,6 +472,12 @@ Libraries for networking
   * Updated:
 
     * The stack size of the MQTT connection monitoring thread can now be adjusted by setting the :kconfig:option:`CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD_STACK_SIZE` Kconfig option.
+    * The library now subscribes to a wildcard cloud-to-device (c2d) topic.
```
After some discussion, we decided the abbreviations for C2D, D2C, and so on should be with capital letters.

jayteemo replied:
the topic string is `"c2d"`, so I would prefer to keep it lower case.
how should i format it? just add quotes?

melwee01 replied:
No quotes or other formatting, just C2D

If you want to keep it lowercase, use ``/c2d`` to refer to the exact topic text.

melwee01 commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -445,6 +472,12 @@ Libraries for networking
 
     * The stack size of the MQTT connection monitoring thread can now be adjusted by setting the :kconfig:option:`CONFIG_NRF_CLOUD_CONNECTION_POLL_THREAD_STACK_SIZE` Kconfig option.
+    * The library now subscribes to a wildcard cloud-to-device (c2d) topic.
+      This enables the device to receive location services data on separate topics.
```
```suggestion
      This enables the device to receive Location Services data on separate topics.
```

Generically referring to 'location data' would also be okay, but Location Services is a capitalized product name.

melwee01 commented on doc/nrf/releases/release-notes-changelog.rst:
```diff
@@ -445,6 +472,12 @@ Libraries for networking
+    * Replaced event :c:enum:`NRF_CLOUD_EVT_RX_DATA` with :c:enum:`NRF_CLOUD_EVT_RX_DATA_GENERAL`.
+    * Added events :c:enum:`NRF_CLOUD_EVT_RX_DATA_CELL_POS` and :c:enum:`NRF_CLOUD_EVT_RX_DATA_SHADOW`.
+    * The library now processes A-GPS and P-GPS data; it is no longer passed to the application.
+    * The status field of :c:enum:`NRF_CLOUD_EVT_ERROR` events use values from the enumeration :c:enumerator:`nrf_cloud_error_status`.
```
```suggestion
    * The status field of :c:enum:`NRF_CLOUD_EVT_ERROR` events uses values from the enumeration :c:enumerator:`nrf_cloud_error_status`.
```

One field, 'of events' doesn't change verb number.

melwee01 commented on samples/nrf9160/nrf_cloud_mqtt_multi_service/README.rst:
```diff
@@ -163,12 +163,7 @@ Location tracking
-The library requests this data automatically, but has no way to automatically receive the response.
-Therefore, all MQTT messages received from `nRF Cloud`_ must be be passed directly through to the :ref:`lib_nrf_cloud_agps` and :ref:`lib_nrf_cloud_pgps` libraries to be checked for A-GPS and P-GPS data, respectively.
-The :c:func:`location_assistance_data_handler` function handles this task, and MQTT messages are sent directly to it from the :file:`src/connection.c` file.
+This amounts to setting up a periodic location request, and then passing the results to a callback configured by the :file:`src/application.c` file.
```
I'd prefer 'involves' to something like 'amounts to'.

melwee01 reviewed (commented):
These are small points, so I don't want to hold up merging, but the branding capitalization definitely needs fixing. 

Perhaps check that there aren't other instances in the same files where Location Services needs to be capitalized.
